mitmproxy/net/http/request.py

```diff
@@ -367,13 +367,64 @@ def _get_urlencoded_form(self):
                 pass
         return ()
 
+    def _sanitized_form_data_replacement(self, form_data):
+        # check if unmodified content changes with URL encoding
+        url_decoded_content = urllib.parse.parse_qsl(self.content, keep_blank_values=True)
+        url_reencoded_content = urllib.parse.urlencode(url_decoded_content, doseq=True).encode()
+        if self.content == url_reencoded_content:
+            # encoding the unmodified content doesn't change anything, so we should be good to go on as usual
+            self.content = mitmproxy.net.http.url.encode(form_data).encode()
+        else:
+            # if we are here, we have to alter encoding behavior to mimic the one used in unmodified POST request
+
+            # check if content isn't encoded at all
```

Comment:
All corner cases not fulfilling the condition `urlencode(urldecode(self.content) == self.content` should enter this `else` tree after early out. So all special cases are handled here. Examples for not well formed self.content:

**1. Keys without value, but no equal sign appended**
```
>>> test_content = b'key_without_value_and_missing_equal_sign&key_without_value_and_missing_equal_sign2'
>>> reencoded_content=urllib.parse.urlencode(urllib.parse.parse_qsl(test_content, keep_blank_values=True), doseq=True).encode()
>>> reencoded_content
b'key_without_value_and_missing_equal_sign=&key_without_value_and_missing_equal_sign2='
>>> reencoded_content == test_content
False
```

To mimic this behavior, equal signs of key without values have to be removed from the result of `mitmproxy.net.http.url.encode(form_data)`

**2. Key, value or both aren't encoded at all in the client request**
```
>>> test_content = b'date[year]=2017'
>>> reencoded_content=urllib.parse.urlencode(urllib.parse.parse_qsl(test_content, keep_blank_values=True), doseq=True).encode()
>>> reencoded_content
b'date%5Byear%5D=2017'
>>> reencoded_content == test_content
False
```

To mimic this behavior, result of `mitmproxy.net.http.url.encode(form_data)` has to be unquoted again.
As there's no urldecode() and parse_qs() / parse_qsl() don't work on `bytes` urllib.parse.unquote has to be used. This should be perfectly fine, as `urllib.parse.urlencode()` uses `quote()` and `quote_plus()` internally, according to [the docs](https://docs.python.org/3/library/urllib.parse.html).

**Other cases which would hit this code:**

```
# space is encoded with quote() instead of quote_plus() by the client sending the request
self.content = b'full%20name=max%20mustermann'

# mixed usage of equal signs on empty keys
self.content = b'key_with_equal_sign=&key_without_equal_sign'
```

and so on...

I think it is clear where this is getting. Most of these examples occur in the POST requests to facebook endpoints which have already been discussed.  Another example has been given [here](https://github.com/mitmproxy/mitmproxy/issues/1946#issuecomment-277522034) by @Kriechi because curl is used with a content-type header 'application/x-www-form-urlencoded' but data is provided without encoding (`--data-urlencode` should be used instead of `--data`). So yes this is somehow wrong tool usage, but if we call it a special case, that would be exactly what we're talking about. Handling some of these cases would make mitmproxy more robust in the end.

What should become clear at this point: There's no way to handle all the cases, this is why I designed the function to act like this:

```
if urlencode(urldecode(self.content) == self.content:
   # go on as usual
else:
   # handle special cases
   if (special case1):
      # modify result of mitmproxy.net.http.url.encode(form_data) accordingly
   elif (special case2):
      # modify result of mitmproxy.net.http.url.encode(form_data) accordingly
   elif (special case3):
      # modify result of mitmproxy.net.http.url.encode(form_data) accordingly
   else:
      # we have some sort of special case which we haven't considered
      # so this is the FAIL OVER for an unhadled case
      # in current implementation fail over is the default behavior of setting:
      # self.content = mitmproxy.net.http.url.encode(form_data).encode()
```